src/ray/object_manager/connection_pool.h

```diff
@@ -0,0 +1,148 @@
+#ifndef RAY_CONNECTION_POOL_H
+#define RAY_CONNECTION_POOL_H
+
+#include <algorithm>
+#include <cstdint>
+#include <deque>
+#include <map>
+#include <memory>
+#include <thread>
+
+#include <boost/asio.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/bind.hpp>
+
+#include "ray/id.h"
+#include "ray/status.h"
+
+#include "format/object_manager_generated.h"
+#include "object_directory.h"
+#include "object_manager_client_connection.h"
+
+namespace asio = boost::asio;
+
+namespace ray {
+
+class ConnectionPool {
+ public:
+  /// Callbacks for GetSender.
+  using SuccessCallback = std::function<void(SenderConnection::pointer)>;
+  using FailureCallback = std::function<void()>;
+
+  /// Connection type to distinguish between message and transfer connections.
+  enum ConnectionType { MESSAGE = 0, TRANSFER };
```

Comment:
`ConnectionType` is used in a lot of signatures. This causes a lot of problems if I remember correctly. However @pcmoritz says that it might work if you do something like `ConnectionType::int` here, is that right?

Reply:
we can use strongly typed enums https://stackoverflow.com/questions/12581064/enum-vs-strongly-typed-enum and specify the type. If we don't and just use ConnectionType, there might be portability problems (I believe we ran into issues in the past).

Reply:
I see. Do you have a type recommendation? In the mean time, I'll make this a strong-typed enum with underlying type `int`.

Reply:
`int` should be fine